File: backend/services/shared/constants/file.go
package constants

import (
	"errors"
	"strings"
)

var ErrOnlyofficeExtensionNotSupported = errors.New("file extension is not supported")

const (
	_OnlyofficeWordType  string = "word"
	_OnlyofficeCellType  string = "cell"
	_OnlyofficeSlideType string = "slide"
)

var OnlyofficeEditableExtensions map[string]string = map[string]string{
	"xlsx": _OnlyofficeCellType,
	"pptx": _OnlyofficeSlideType,
	"docx": _OnlyofficeWordType,
}

var OnlyofficeFileExtensions map[string]string = map[string]string{
	"xls":  _OnlyofficeCellType,
	"xlsx": _OnlyofficeCellType,
	"xlsm": _OnlyofficeCellType,
	"xlt":  _OnlyofficeCellType,
	"xltx": _OnlyofficeCellType,
	"xltm": _OnlyofficeCellType,
	"ods":  _OnlyofficeCellType,
	"fods": _OnlyofficeCellType,
	"ots":  _OnlyofficeCellType,
	"csv":  _OnlyofficeCellType,
	"pps":  _OnlyofficeSlideType,
	"ppsx": _OnlyofficeSlideType,
	"ppsm": _OnlyofficeSlideType,
	"ppt":  _OnlyofficeSlideType,
	"pptx": _OnlyofficeSlideType,
	"pptm": _OnlyofficeSlideType,
	"pot":  _OnlyofficeSlideType,
	"potx": _OnlyofficeSlideType,
	"potm": _OnlyofficeSlideType,
	"odp":  _OnlyofficeSlideType,
	"fodp": _OnlyofficeSlideType,
	"otp":  _OnlyofficeSlideType,
	"doc":  _OnlyofficeWordType,
	"docx": _OnlyofficeWordType,
	"docm": _OnlyofficeWordType,
	"dot":  _OnlyofficeWordType,
	"dotx": _OnlyofficeWordType,
	"dotm": _OnlyofficeWordType,
	"odt":  _OnlyofficeWordType,
	"fodt": _OnlyofficeWordType,
	"ott":  _OnlyofficeWordType,
	"rtf":  _OnlyofficeWordType,
	"txt":  _OnlyofficeWordType,
	"html": _OnlyofficeWordType,
	"htm":  _OnlyofficeWordType,
	"mht":  _OnlyofficeWordType,
	"pdf":  _OnlyofficeWordType,
	"djvu": _OnlyofficeWordType,
	"fb2":  _OnlyofficeWordType,
	"epub": _OnlyofficeWordType,
	"xps":  _OnlyofficeWordType,
}

func IsExtensionSupported(fileExt string) bool {
	_, exists := OnlyofficeFileExtensions[strings.ToLower(fileExt)]
	if exists {
		return true
	}
	return false
}

func IsExtensionEditable(fileExt string) bool {
	_, exists := OnlyofficeEditableExtensions[strings.ToLower(fileExt)]
	if exists {
		return true
	}
	return false
}

func GetFileType(fileExt string) (string, error) {
	fileType, exists := OnlyofficeFileExtensions[strings.ToLower(fileExt)]
	if !exists {
		return "", ErrOnlyofficeExtensionNotSupported
	}
	return fileType, nil
}

File: frontend/src/utils/file.ts
import docx from "@assets/docx.svg";
import pptx from "@assets/pptx.svg";
import xls from "@assets/xls.svg";
import notsupported from "@assets/nofile.svg";

const DOCUMENT_EXTS = [
  "doc",
  "docx",
  "docm",
  "dot",
  "dotx",
  "dotm",
  "odt",
  "fodt",
  "ott",
  "rtf",
  "txt",
  "html",
  "htm",
  "mht",
  "xml",
  "pdf",
  "djvu",
  "fb2",
  "epub",
  "xps",
  "oxps",
];
const SPREADSHEET_EXTS = [
  "xls",
  "xlsx",
  "xlsm",
  "xlt",
  "xltx",
  "xltm",
  "ods",
  "fods",
  "ots",
  "csv",
];
const PRESENTATION_EXTS = [
  "pps",
  "ppsx",
  "ppsm",
  "ppt",
  "pptx",
  "pptm",
  "pot",
  "potx",
  "potm",
  "odp",
  "fodp",
  "otp",
];

const EDITABLE_EXTS = ["docx", "pptx", "xlsx"];
const OPENABLE_EXTS =
  DOCUMENT_EXTS.concat(SPREADSHEET_EXTS).concat(PRESENTATION_EXTS);

const WORD = "word";
const SLIDE = "slide";
const CELL = "cell";

const getFileExt = (filename: string): string =>
  filename.split(".").pop() || "";

export const isFileEditable = (filename: string) => {
  const ext = getFileExt(filename).toLowerCase();
  return EDITABLE_EXTS.includes(ext);
};

export const isFileSupported = (filename: string) => {
  const e = getFileExt(filename).toLowerCase();
  return OPENABLE_EXTS.includes(e);
};

export const getFileType = (filename: string) => {
  const e = getFileExt(filename).toLowerCase();

  if (DOCUMENT_EXTS.includes(e)) return WORD;
  if (SPREADSHEET_EXTS.includes(e)) return CELL;
  if (PRESENTATION_EXTS.includes(e)) return SLIDE;

  return null;
};

export const getFileIcon = (filename: string) => {
  const e = getFileExt(filename).toLowerCase();

  if (DOCUMENT_EXTS.includes(e)) return docx;
  if (SPREADSHEET_EXTS.includes(e)) return xls;
  if (PRESENTATION_EXTS.includes(e)) return pptx;

  return notsupported;
};

// TODO: Set proper defaults
export const getCreateFileUrl = (
  fileType: "docx" | "pptx" | "xlsx" | undefined
) => {
  switch (fileType) {
    case "docx":
      return encodeURIComponent(process.env.WORD_FILE || "");
    case "pptx":
      return encodeURIComponent(process.env.SLIDE_FILE || "");
    case "xlsx":
      return encodeURIComponent(process.env.SPREADSHEET_FILE || "");
    default:
      return encodeURIComponent(process.env.WORD_FILE || "");
  }
};

export const formatBytes = (bytes: number, decimals = 2) => {
  if (!+bytes) return "0 Bytes";

  const k = 1024;
  const dm = decimals < 0 ? 0 : decimals;
  const sizes = ["Bytes", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB"];

  const i = Math.floor(Math.log(bytes) / Math.log(k));

  return `${parseFloat((bytes / k ** i).toFixed(dm))} ${sizes[i]}`;
};
